torch/csrc/jit/passes/autocast.cpp

```diff
@@ -0,0 +1,456 @@
+      for (Use use : value->uses()) {
+        // TODO: support runtime flag
+        if (use.user->kind() == prim::SetAttr &&
+            use.user->s(attr::name) == "_enabled") {
+          // Search for `prim::SetAttr[name="_enabled"]`
+          auto ret = constant_as<bool>(use.user->input(1));
+          TORCH_CHECK(
+              ret.has_value(), "Autocast _enabled argument must be a constant");
+          enabled = ret.value();
+        } else if (
+            use.user->kind() == prim::SetAttr &&
+            use.user->s(attr::name) == "device") {
+          // Search for `prim::SetAttr[name="device"]`
+          auto ret = constant_as<std::string>(use.user->input(1));
+          TORCH_CHECK(
+              ret.has_value(), "Autocast device argument must be a constant");
+          device = ret.value();
+        } else if (
+            use.user->kind() == prim::SetAttr &&
+            use.user->s(attr::name) == "fast_dtype") {
+          // Search for `prim::SetAttr[name="fast_dtype"]`
+          auto ret = constant_as<c10::ScalarType>(use.user->input(1));
+          TORCH_CHECK(
+              ret.has_value() && ret.value() != c10::ScalarType::Undefined,
+              "Autocast dtype argument must be a constant and defined");
+          dtype = ret.value();
+        }
+      }
+      TORCH_CHECK(enabled.has_value(), "Autocast missing _enabled attribute");
+      TORCH_CHECK(
+          dtype != c10::ScalarType::Undefined,
+          "Autocast missing fast_dtype attribute");
+      TORCH_CHECK(!device.empty(), "Autocast missing device attribute");
+      if (device == "cuda") {
+        scope.context.enabled = enabled.value();
+        scope.context.scalar_type = dtype;
+      } else if (device == "cpu") {
+        scope.context.cpu_enabled = enabled.value();
+        scope.context.cpu_scalar_type = dtype;
+      } else {
+        TORCH_INTERNAL_ASSERT(
+            false, "unrecognized device for autocast pass: ", device);
+      }
+      return scope;
+    } else {
+      // We only support simple and static autocast expressions. For example,
+      // the following should report an error (since the autocast would not
+      // work as expected)
+      //
+      //    autocast_on = autocast(enabled=True)
+      //    autocast_off = autocast(enabled=False)
+      //    with autocast_on if condition else autocast_off:
+      //        ...
+      //
+      // TODO: better error message
+      //
+      AT_ERROR("Unsupported autocast syntax");
+    }
+  }
+
+  // Not an autocast...
+  return c10::nullopt;
+}
+
+void castTensorInputs(
+    Node* node,
+    Symbol cast_op,
+    const AutocastContext& context) {
+  if (!context) {
+    return;
+  }
+
+  const auto graph = node->owningGraph();
+
+  std::unordered_set<Value*> casted_inputs;
+  for (auto input : node->inputs()) {
+    // TODO: update cast_op signature to take dynamic context flags
+    auto input_tensor_type = input->type()->cast<TensorType>();
```

Comment:
do we need to support Optional inputs like conv2d bias ? We might need overload of op that takes in Optional Tensor and returns Optional Tensor

Reply:
Good catch, we would need that! Retrieving the proper scalar_type might be trickier for Optional Tensor as well, but it's not properly handled right now neither. :stuck_out_tongue_closed_eyes: